src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -458,44 +472,36 @@ void AtlasEngine::_present()
         return;
     }
 
-    if constexpr (!ATLAS_DEBUG_SHOW_DIRTY)
+    if (!ATLAS_DEBUG_SHOW_DIRTY && !_p.s->target->disablePresent1 && memcmp(&dirtyRect, &fullRect, sizeof(RECT)) != 0)
     {
-        if (memcmp(&dirtyRect, &fullRect, sizeof(RECT)) != 0)
-        {
-            params.DirtyRectsCount = 1;
-            params.pDirtyRects = &dirtyRect;
+        params.DirtyRectsCount = 1;
+        params.pDirtyRects = &dirtyRect;
 
-            if (_p.scrollOffset)
-            {
-                const auto offsetInPx = _p.scrollOffset * _p.s->font->cellSize.y;
-                const auto width = _p.s->targetSize.x;
-                // We don't use targetSize.y here, because "height" refers to the bottom coordinate of the last text row
-                // in the buffer. We then add the "offsetInPx" (which is negative when scrolling text upwards) and thus
-                // end up with a "bottom" value that is the bottom of the last row of text that we haven't invalidated.
-                const auto height = _p.s->viewportCellCount.y * _p.s->font->cellSize.y;
-                const auto top = std::max(0, offsetInPx);
-                const auto bottom = height + std::min(0, offsetInPx);
-
-                scrollRect = { 0, top, width, bottom };
-                scrollOffset = { 0, offsetInPx };
-
-                params.pScrollRect = &scrollRect;
-                params.pScrollOffset = &scrollOffset;
-            }
-        }
-    }
-
-    if constexpr (Feature_AtlasEnginePresentFallback::IsEnabled())
-    {
-        if (FAILED_LOG(_p.swapChain.swapChain->Present1(1, 0, &params)))
+        if (_p.scrollOffset)
         {
-            THROW_IF_FAILED(_p.swapChain.swapChain->Present(1, 0));
+            const auto offsetInPx = _p.scrollOffset * _p.s->font->cellSize.y;
+            const auto width = _p.s->targetSize.x;
+            // We don't use targetSize.y here, because "height" refers to the bottom coordinate of the last text row
+            // in the buffer. We then add the "offsetInPx" (which is negative when scrolling text upwards) and thus
+            // end up with a "bottom" value that is the bottom of the last row of text that we haven't invalidated.
+            const auto height = _p.s->viewportCellCount.y * _p.s->font->cellSize.y;
+            const auto top = std::max(0, offsetInPx);
+            const auto bottom = height + std::min(0, offsetInPx);
+
+            scrollRect = { 0, top, width, bottom };
+            scrollOffset = { 0, offsetInPx };
+
+            params.pScrollRect = &scrollRect;
+            params.pScrollOffset = &scrollOffset;
         }
     }
-    else
+
+    auto hr = _p.swapChain.swapChain->Present1(1, 0, &params);
+    if (Feature_AtlasEnginePresentFallback::IsEnabled() && FAILED(hr) && params.DirtyRectsCount != 0)

```

Comment:
FWIW, Velocity checks are (by Windows style) usually in their own conditionals, even if it makes the code more annoying to read. However, eh.